First stab at sponsored transaction precompile (#3074)
ethereum/EIPs: EIPS/eip-3074.md

 - **`NONCE_PRECOMPILE`** - the specific precompile at address `0x14`, introduced by this EIP, which exposes relayer-sponsee nonces.
 - **Transaction-like Package** - the signed arguments passed to `CALL_PRECOMPILE`.
 - **Sponsor** - the account which is responsible for paying gas fees and sending the transaction. May or may not be the same as the relayer.
 - **Sponsee** - the account which signed the transactions-like package.
 - **Relayer** - the account or contract which directly calls into `CALL_PRECOMPILE`. May or may not be the same as the sponsor.
 - **Callee** - the target of the call from `CALL_PRECOMPILE`.

### API

#### `CALL_PRECOMPILE`

`CALL_PRECOMPILE` requires the following eight arguments:

 - `nonce` - the next nonce value, as described below;
 - `to` - the address of the callee (not `CALL_PRECOMPILE`);
 - `gaslimit` - the minimum gas limit which must be provided with the call into `CALL_PRECOMPILE`;
 - `value` - the exact amount of Ether in wei to transfer from the relayer to the callee;
 - `data` - the calldata for the call into `to`; and

>  - `v`, `r`, `s` - signature for the package, including chain id as specified in [EIP-115](./eip-115.md).

Reviewer: ```suggestion
 - `v`, `r`, `s` - signature for the package, including chain id as specified in [EIP-155](./eip-155.md).
```
I'm pretty sure this is what you meant.  Thank God for the EIP validator that caught this (it is what is blocking CI from letting me merge)!

Assuming the calldata is the correct length, `NONCE_PRECOMPILE` will return a success, and place the nonce associated with the address pair in the return data.

### Nonces

The two precompiles will maintain a nonce for each pair of relayer address and sponsee address, in essence:

```
{
    (0x1234...5678, 0xEEEE...EEEE) => 55,
    (0x1122...3344, 0xBBBB...BBBB) => 89,
}
```

Where:
 * `0x1234...5678` and `0x1122...3344` are the relayer addresses;
 * `0xEEEE...EEEE` and `0xBBBB...BBBB` are the sponsee addresses; and
 * `55` and `89` are the current nonce values for their respective pairs.

The nonce shall be incremented whenever a correctly signed transaction-like package containing the next nonce is submitted to `CALL_PRECOMPILE` with a sufficient gas limit.

### Gas Fees

#### `CALL_PRECOMPILE`

TODO: Probably something like the sum of:
 * The cost of a normal call, including calldata and signature size, etc.
 * An `SLOAD` to read the current nonce
 * An `SSTORE` to write the updated nonce
 * Cost of an ecrecover

#### `NONCE_PRECOMPILE`

TODO: Probably something like the sum of:
 * An `SLOAD` to read the current nonce

## Rationale

### Another Sponsored Transaction EIP

> Other approaches to sponsored transactions, like [EIP-2711](./eip-2711.md) and [EIP-2733](./eip-2733.md), are incompatible with [EIP-2938](./eip-2938.md) flavored account abstraction (AA). These proposals require a _signed_ transaction from the sponsor's account, which is not possible from an AA contract, because it has no private key to sign with.

Reviewer: You may want to consider not including references to other competing draft EIPs in this EIP as it creates a dependency that you probably don't want.  It would be unfortunate if this EIP wasn't able to go to final because it points at draft EIPs internally!  Instead, consider just *briefly* describing the alternatives (like a sentence or two) inline and then describing why this alternative is better.  For example:
> We could have done XXX but that would not work with any future account abstraction proposals.

---
eip: 3074
title: Sponsored Transaction Precompile
author: Sam Wilson (@SamWilsn)

> discussions-to: <URL>

Reviewer: This is required for draft EIPs.  We recommend a forum post at https://ethereum-magicians.org/, but you can also create a GitHub issue for discussion if you prefer.  You cannot use this PR as discussion-to URL.

status: Draft

> type: Standard

Reviewer 2: ```suggestion
type: Standards Track
```

category: Core
created: 2020-10-15
---

## Simple Summary

> Creates a new precompile, analogous to `CALL`, that sets `msg.sender` based on an ECDSA signature.

Reviewer 2: ```suggestion
Creates a new precompile, analogous to `CALL (0xF1)`, that sets `CALLER (0x33)` based on an ECDSA signature.
```

## Abstract

We propose adding two precompiles:

>  - `CALL_PRECOMPILE` - forwards a `CALL`, setting `msg.sender` according to an ECDSA signature, using a relayer-sponsee nonce for replay protection.

Reviewer 2: ```suggestion
 - `CALL_PRECOMPILE` - forwards a `CALL`, setting `CALLER` according to an ECDSA signature, using a relayer-sponsee nonce for replay protection.
```

Reviewer 2: Just a nit, but I generally prefer to not mix EVM and Solidity syntax. Feel free to disregard.

 - `NONCE_PRECOMPILE` - provides access to relayer-sponsee nonces expected by `CALL_PRECOMPILE`.

## Motivation

Sponsored transactions—the separation of fee payment from transaction content—have been a long standing feature request. Unlike similar proposals, this EIP specifies a method of implementing sponsored transactions that allows both externally owned accounts (EOAs) and [EIP-2938] contracts to act as sponsors.

With the explosion of tokens built on Ethereum, especially stable coins, it has become common for EOAs to hold valuable assets without holding any Ether at all. These assets must be converted to Ether before they can be used to pay gas fees, but without Ether to pay for the conversion, it's impossible to convert them. Sponsored transactions break the circular dependency.

While it is possible to emulate sponsored transactions (ex. [Gas Station Network][gsn]), these solutions require specific support in callee contracts.

## Specification

> We propose adding two precompiles to implement sponsored transactions:

Reviewer: EIPs don't propose, they assert.  The target audience for an EIP is people 5 to 10 years in the future, after this EIP has been final for years.

 - `sponsee` - the sponsee address.

Assuming the calldata is the correct length, `NONCE_PRECOMPILE` will return a success, and place the nonce associated with the address pair in the return data.

### Nonces

The two precompiles will maintain a nonce for each pair of relayer address and sponsee address, in essence:

```
{
    (0x1234...5678, 0xEEEE...EEEE) => 55,
    (0x1122...3344, 0xBBBB...BBBB) => 89,
}
```

Where:
 * `0x1234...5678` and `0x1122...3344` are the relayer addresses;
 * `0xEEEE...EEEE` and `0xBBBB...BBBB` are the sponsee addresses; and
 * `55` and `89` are the current nonce values for their respective pairs.

The nonce shall be incremented whenever a correctly signed transaction-like package containing the next nonce is submitted to `CALL_PRECOMPILE` with a sufficient gas limit.

### Gas Fees

#### `CALL_PRECOMPILE`

TODO: Probably something like the sum of:
 * The cost of a normal call, including calldata and signature size, etc.
 * An `SLOAD` to read the current nonce
 * An `SSTORE` to write the updated nonce
 * Cost of an ecrecover

#### `NONCE_PRECOMPILE`

TODO: Probably something like the sum of:
 * An `SLOAD` to read the current nonce

## Rationale

> 
> <!-- 
> The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.
> -->

Reviewer: ```suggestion
```

More logic can be implemented around the call into `CALL_PRECOMPILE`, giving more control to relayers and sponsors without sacrificing security or user experience for sponsees.

### Nonces

Other nonce schemes either do not provide enough security, or are too inefficient/inconvenient to be practical.

 - Use sponsor nonce: every transaction from the sponsor's account invalidates every transaction-like package.
 - Use relayer nonce: every sponsored transaction from the relayer's account invalidates every other transaction-like package. Also interacts with `SELFDESTRUCT`.
 - Use sponsee nonce: the sponsee could attack the sponsor by submitting another transaction with a conflicting nonce at a higher gas price.

Instead, by creating an independent nonce per relayer-sponsee pair, we get some attractive properties:
 - A transaction package can only be invalidated if both the relayer and sponsee cooperate, which is nice for EOA sponsors, and necessary for AA contracts.
 - The `SELFDESTRUCT` operation doesn't introduce replay attacks.

## Backwards Compatibility

No known issues.

## Test Cases

TODO

## Implementation

> 
> TODO
> 
> <!--
> The implementations must be completed before any EIP is given status "Final", but it need not be completed before the EIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.
> -->

Reviewer: ```suggestion

TODO
```

## Security Considerations

 - First precompiles that require persistent storage.
 - Potential impersonation attacks if there is a bug in the signature verification.
 - Potential replay-attack problems if there is a bug in the replay protection, or if two chains share chain ids.

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

> 
> ## Acknowledgments
> 
> Largely inspired by [EIP ? - CallWithSigner as a potential fix for the msg.sender problem][EIP-?] and [EIP-2711].

Reviewer: ```suggestion
```

As technical documents, props, acknowledgments, thanks, etc. are generally frowned on.  You can include these in the discussion-to link, or in an associated blog post, etc. if you like but they aren't really appropriate for a technical specification (especially a public domain one).

[EIP-2711]: ./eip-2711.md
[EIP-2733]: ./eip-2733.md
[EIP-2938]: ./eip-2938.md
[EIP-115]: ./eip-155.md
[EIP-?]: https://ethereum-magicians.org/t/eip-callwithsigner-as-a-potential-fix-for-the-msg-sender-problem/4340

> [gsn]: https://www.opengsn.org/

Reviewer: Can these be inlined instead of included as references?  Most of these links are quite short, and I think the one long one will be removed with the removal of acknowledgements.  This would be for repository wide stylistic reasons more than anything, as the *vast* majority of EIPs just inline external links.
